src/gt4py/next/iterator/type_system/inference.py

```diff
@@ -102,7 +102,9 @@ def copy_type(from_: itir.Node, to: itir.Node, allow_untyped=False) -> None:
     This function mainly exists for readability reasons.
     """
     assert allow_untyped is not None or isinstance(from_.type, ts.TypeSpec)
-    _set_node_type(to, from_.type)  # type: ignore[arg-type]
+    if from_.type is None:
+        return
```

Comment:
> Does that mean, that if from_ has no type, the type of to will be kept?
Yes, this is the intend.
> Good catch, this might not be desired in the default case. I changed it to be only allowed when `allow_untyped` is `True`.